textattack/search_methods/particle_swarm_optimization.py

```diff
@@ -24,252 +25,313 @@ class ParticleSwarmOptimization(SearchMethod):
+            for k in range(len(population)):
+                change_ratio = count_change_ratio(
+                    population[k], initial_result.attacked_text
                 )
-                p_change = (
-                    1 - 2 * change_ratio
-                )  # referred from the original source code
+                # Referred from the original source code
+                p_change = 1 - 2 * change_ratio
                 if np.random.uniform() < p_change:
-                    new_h, new_w_list = self._gen_h_score(
-                        x, neighbors_len, neighbors_list
-                    )
-                    new_pop.append(self._mutate(x, new_h, new_w_list))
-                else:
-                    new_pop.append(x)
-            pop = new_pop
-
-            # check if there is any successful attack in the current population
-            pop_results, self.search_over = self.get_goal_results(pop)
-            if self.search_over:
-                return max(pop_results, key=lambda x: x.score)
-            pop_scores = np.array([r.score for r in pop_results])
-            top_attack = np.argmax(pop_scores)
-            if (
-                pop_results[top_attack].goal_status
-                == GoalFunctionResultStatus.SUCCEEDED
-            ):
-                return pop_results[top_attack]
-
-            # update the elite if the score is increased
-            for k in range(self.pop_size):
-                if pop_scores[k] > part_elites_scores[k]:
-                    part_elites[k] = pop[k]
-                    part_elites_scores[k] = pop_scores[k]
-            if pop_scores[top_attack] > all_elite_score:
-                all_elite = pop[top_attack]
-                all_elite_score = pop_scores[top_attack]
-
-        return initial_result
+                    self._perturb(population[k], initial_result)
+
+                if self._search_over:
+                    break
+            if self._search_over:
+                break
+
+            # Check if there is any successful attack in the current population
+            top_result = max(population, key=lambda x: x.score)
+            if top_result.result.goal_status == GoalFunctionResultStatus.SUCCEEDED:
+                return top_result.result
+
+            # Update the elite if the score is increased
+            for k in range(len(population)):
+                if population[k].score > local_elites[k].score:
+                    local_elites[k] = deepcopy(population[k])
+
+            if top_result.score > global_elite.score:
+                global_elite = deepcopy(top_result)
+
+        return global_elite.result
+
+    def check_transformation_compatibility(self, transformation):
+        """The genetic algorithm is specifically designed for word
+        substitutions."""
+        return transformation_consists_of_word_swaps(transformation)
+
+    def extra_repr_keys(self):
+        return ["pop_size", "max_iters", "post_turn_check", "max_turn_retries"]
+
+
```

Comment:
Yes we can move `sigmoid` to `utils` and `count_change_ratio` to `AttackedText`, but I would just keep `normalize` where it is since normalization can depend on the specific method.